bpo-28254: Add a C-API for controlling the state of the garbage collector (#25687)
python/cpython: Doc/c-api/gcsupport.rst

   this method (don't just call :c:func:`Py_DECREF` on a reference).  The
   collector will call this method if it detects that this object is involved
   in a reference cycle.

Controlling the Garbage Collector State
---------------------------------------

The C-API provides the following functions for controlling
garbage collection runs.

> 

Reviewer: Once you writes "Performs" and once "Enable" (no S). I suggest to avoid the S, most documentations avoid it.

.. c:function:: Py_ssize_t PyGC_Collect(void)

>    Performs a garbage collection, if the garbage collector is enabled.

Reviewer: ```suggestion
   Performs a full garbage collection, if the garbage collector is enabled: same as calling func:`gc.collect`.
```

You may mention that free lists of built-in types are also cleared.

Author: It's actually not the same as `gc.collect()`. The latter doesn't care about `gc.enabled`, whereas `PyGC_Collect()` honours it.

Reviewer: > I don't think this C-API function is the right place to document these details. That's what the gc module is for.

I agree that gc.collect() documentation must be the most complete. But in this case, mention that PyGC_Collect is similar to gc.collect() but just document changes between PyGC_Collect() and gc.collect(). As you wrote, the main one is that PyGC_Collect() honors gc.enabled.

Reviewer 2: I can document `PyGC_Collect` later, let's focus this PR on the new APIs.

Reviewer: I'm fine with documenting this "complex" function in a separated PR.

Author: I added a sentence to compare the two. Feel free to take another pass later.

>    Returns the number of collected + uncollectable objects.

Reviewer: Hum, uncollectable objects are stored in gc.garbage no? Shoud we document that?

The C code uses "unreachable objects that couldn't be collected" in a comment instead of "uncollectable objects". Maybe it's a more explicit definition.

Author: I don't think this C-API function is the right place to document these details. That's what the `gc` module is for.

>    If the garbage collector is disabled or already running,
>    returns ``0`` immediately.

Reviewer: ```suggestion

   If the garbage collector is disabled or already collecting,
   does nothing and returns ``0`` immediately.
   
```

>    Errors during garbage collection are ignored and printed.

Reviewer: Errors cannot be ignored and printed at the same time :-) Exceptions are logged with sys.unraisablehook, you can add a link using:

```
:data:`sys.unraisablehook`
```

Author: Done.

>    This function does not raise exceptions.

Reviewer: @pablogsal: Do we want to document that it's safe to call this function with an exception raised, and that the exception is saved and then restored?

   this method (don't just call :c:func:`Py_DECREF` on a reference).  The
   collector will call this method if it detects that this object is involved
   in a reference cycle.

Controlling the Garbage Collector State
---------------------------------------

The C-API provides the following functions for controlling
garbage collection runs.

.. c:function:: Py_ssize_t PyGC_Collect(void)

   Perform a full garbage collection, if the garbage collector is enabled.
   Returns the number of collected + unreachable objects which cannot
   be collected.
   If the garbage collector is disabled or already collecting,
   returns ``0`` immediately.
   Errors during garbage collection are passed to :data:`sys.unraisablehook`.
   This function does not raise exceptions.

.. c:function:: int PyGC_Enable(void)

>    Enable the garbage collector.

Reviewer: ```suggestion
   Enable the garbage collector: similar to :func:`gc.enable`.
```

>    Returns the previous state, 0 for disabled and 1 for enabled.
> 
> .. versionchanged:: 3.10

Reviewer: Oh, the markup should be at the same indentation level than the paragraph. Also "versionadded" is better than "versionchanged".

```suggestion
   Returns the previous state, 0 for disabled and 1 for enabled.

   .. versionadded:: 3.10
```

Same remark for the 2 other added functions.

Author: Right. I had thought of it and then still ended up copynpasting… :-/

.. c:function:: int PyGC_Disable(void)

>    Disable the garbage collector.

Reviewer: ```suggestion
   Disable the garbage collector: similar to :func:`gc.disable`.
```

   Returns the previous state, 0 for disabled and 1 for enabled.

.. versionchanged:: 3.10

.. c:function:: int PyGC_IsEnabled(void)

>    Query the state of the garbage collector.

Reviewer: ```suggestion
   Query the state of the garbage collector: similar to :func:`gc.isenabled`.
```

   Enable the garbage collector: similar to :func:`gc.enable`.
   Returns the previous state, 0 for disabled and 1 for enabled.

> .. versionadded:: 3.10

Reviewer: Please fix also the indentation.
